# HT 2 : Сжатие модели DistilBERT
---
Задача: уменьшить размер модели в 2 раза с минимальной потерей качества

### 🔧 Подготовка среды

In [4]:
# !pip3 install torch torchvision transformers

### 1️⃣ Импорт библиотек и загрузка модели

In [2]:

from transformers import AutoModelForSequenceClassification, AutoTokenizer, pipeline
from datasets import load_dataset
import torch, os

MODEL_ID = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
model = AutoModelForSequenceClassification.from_pretrained(MODEL_ID)

dataset = load_dataset("imdb", split="test", )


### 2️⃣ Проверка работы исходной модели

In [3]:
clf = pipeline("sentiment-analysis", model=model, tokenizer=tokenizer)

texts = [
    "I loved this movie, it was fantastic!",
    "This was a terrible waste of time.",
    "Not bad, but could be better.",
]

for t in texts:
    print(f"\nText: {t}")
    print("Prediction:", clf(t))


Device set to use cpu



Text: I loved this movie, it was fantastic!
Prediction: [{'label': 'POSITIVE', 'score': 0.9998804330825806}]

Text: This was a terrible waste of time.
Prediction: [{'label': 'NEGATIVE', 'score': 0.9998123049736023}]

Text: Not bad, but could be better.
Prediction: [{'label': 'NEGATIVE', 'score': 0.8015893697738647}]


### 3️⃣ Замер размера исходной модели

In [4]:

torch.save(model.state_dict(), "model_fp32.pt")
size_fp32 = os.path.getsize("model_fp32.pt") / 1e6
print(f" Размер исходной модели: {size_fp32:.1f} MB")


 Размер исходной модели: 267.9 MB


### 4️⃣ Квантование

In [5]:
from hw2 import *

In [6]:
quantized_model = quantize_model(model, modules_to_quantize = {nn.Linear})
size_quantized = save_model(quantized_model, "model_quantized.pt")

print(f"Размер квантованной модели: {size_quantized:.1f} MB")
print(f"Уменьшение размера: {(1 - size_quantized/size_fp32)*100:.1f}%")

Размер квантованной модели: 138.7 MB
Уменьшение размера: 48.2%


### 5️⃣ Оценка качества

In [7]:
print("Evaluating original model...")
original_metrics = evaluate_model_quality(
    model,
    tokenizer,
    dataset,
    max_samples=1000,
    batch_size=32
)

print("\nOriginal Model Metrics:")
for key, value in original_metrics.items():
    if isinstance(value, float):
        print(f"  {key}: {value:.4f}")
    else:
        print(f"  {key}: {value}")

Evaluating original model...


Evaluating: 100%|█████████████████████████████████████████████████████| 32/32 [05:03<00:00,  9.47s/it]


Original Model Metrics:
  accuracy: 0.9080
  total_samples: 1000
  correct_predictions: 908


In [19]:

def evaluate_model_quality(
    model: nn.Module,
    tokenizer,
    validation_dataset,
    max_samples: int = 1000,
    batch_size: int = 32,
    max_length: int = 512
) -> Dict:
    """
    Args:
        model: Model to evaluate
        tokenizer: Tokenizer for the model
        validation_dataset: Dataset with 'text' and 'label' fields
        max_samples: Maximum number of samples to evaluate
        batch_size: Batch size for evaluation
        max_length: Maximum sequence length
    
    Returns:
        Dictionary with accuracy and other metrics
    """
    model.eval()
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)
    
    correct = 0
    total = 0
    all_predictions = []
    all_labels = []
    eval_samples = min(max_samples, len(validation_dataset))
    
    with torch.no_grad():
        for i in tqdm(range(0, eval_samples, batch_size), desc="Evaluating"):
            batch_end = min(i + batch_size, eval_samples)
            batch = validation_dataset[i:batch_end]
            
            inputs = tokenizer(
                batch['text'],
                padding=True,
                truncation=True,
                max_length=max_length,
                return_tensors="pt"
            )
            inputs = {k: v.to(device) for k, v in inputs.items()}
            labels = torch.tensor(batch['label']).to(device)
            
            outputs = model(**inputs)
            predictions = torch.argmax(outputs.logits, dim=-1)
            
            correct += (predictions == labels).sum().item()
            total += labels.size(0)
            
            all_predictions.extend(predictions.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())
    
    accuracy = correct / total
    
    # Calculate additional metrics
    all_predictions = np.array(all_predictions)
    all_labels = np.array(all_labels)

    print(np.unique(all_labels))
    
    # For binary classification
    if len(np.unique(all_labels)) == 2:
        true_positives = np.sum((all_predictions == 1) & (all_labels == 1))
        false_positives = np.sum((all_predictions == 1) & (all_labels == 0))
        false_negatives = np.sum((all_predictions == 0) & (all_labels == 1))
        
        precision = true_positives / (true_positives + false_positives) if (true_positives + false_positives) > 0 else 0
        recall = true_positives / (true_positives + false_negatives) if (true_positives + false_negatives) > 0 else 0
        f1_score = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
    else:
        precision = recall = f1_score = None
    
    metrics = {
        'accuracy': accuracy,
        'total_samples': total,
        'correct_predictions': correct
    }
    
    if precision is not None:
        metrics.update({
            'precision': precision,
            'recall': recall,
            'f1_score': f1_score
        })
    
    return metrics



In [20]:
print("Evaluating original model...")
original_metrics = evaluate_model_quality(
    model,
    tokenizer,
    dataset,
    max_samples=64,
    batch_size=32
)

print("\nOriginal Model Metrics:")
for key, value in original_metrics.items():
    if isinstance(value, float):
        print(f"  {key}: {value:.4f}")
    else:
        print(f"  {key}: {value}")

Evaluating original model...


Evaluating: 100%|███████████████████████████████████████████████████████| 2/2 [00:11<00:00,  5.71s/it]

[0]

Original Model Metrics:
  accuracy: 0.8750
  total_samples: 64
  correct_predictions: 56


In [8]:
print("Evaluating quantized model...")
quantized_metrics = evaluate_model_quality(
    quantized_model,
    tokenizer,
    dataset,
    max_samples=1000,
    batch_size=32
)

print("\nQuantized Model Metrics:")
for key, value in quantized_metrics.items():
    if isinstance(value, float):
        print(f"  {key}: {value:.4f}")
    else:
        print(f"  {key}: {value}")

Evaluating quantized model...


Evaluating: 100%|█████████████████████████████████████████████████████| 32/32 [04:05<00:00,  7.66s/it]


Quantized Model Metrics:
  accuracy: 0.8950
  total_samples: 1000
  correct_predictions: 895


In [9]:
# Compare accuracy
accuracy_drop = (original_metrics['accuracy'] - quantized_metrics['accuracy']) * 100
print(f"\nAccuracy drop: {accuracy_drop:.2f}%")

if 'f1_score' in original_metrics:
    f1_drop = (original_metrics['f1_score'] - quantized_metrics['f1_score']) * 100
    print(f"F1 Score drop: {f1_drop:.2f}%")


Accuracy drop: 1.30%


### 6️⃣ (Дополнительно) Проверка скорости инференса

In [10]:
print("Measuring inference time for original model...")
original_time = measure_inference_time(
    model,
    tokenizer,
    texts,
    num_runs=100
)

print("\n⚡ Original Model Inference Time:")
print(f"  Mean: {original_time['mean_time']*1000:.2f} ms")
print(f"  Std:  {original_time['std_time']*1000:.2f} ms")

Measuring inference time for original model...


Measuring inference time: 100%|█████████████████████████████████████| 100/100 [00:04<00:00, 20.12it/s]


⚡ Original Model Inference Time:
  Mean: 15.93 ms
  Std:  0.54 ms


In [11]:
print("Measuring inference time for quantized model...")
quantized_time = measure_inference_time(
    quantized_model,
    tokenizer,
    texts,
    num_runs=100
)

print("\n⚡ Quantized Model Inference Time:")
print(f"  Mean: {quantized_time['mean_time']*1000:.2f} ms")
print(f"  Std:  {quantized_time['std_time']*1000:.2f} ms")

Measuring inference time for quantized model...


Measuring inference time: 100%|█████████████████████████████████████| 100/100 [00:02<00:00, 34.41it/s]


⚡ Quantized Model Inference Time:
  Mean: 9.24 ms
  Std:  2.47 ms


In [12]:
speedup = original_time['mean_time'] / quantized_time['mean_time']
print(f"\n🚀 Speedup: {speedup:.2f}x")


🚀 Speedup: 1.72x


### Проверка работы квантованной модели

In [13]:
clf = pipeline("sentiment-analysis", model=quantized_model, tokenizer=tokenizer)

for t in texts:
    print(f"\nText: {t}")
    print("Prediction:", clf(t))

Device set to use cpu



Text: I loved this movie, it was fantastic!
Prediction: [{'label': 'POSITIVE', 'score': 0.9998642206192017}]

Text: This was a terrible waste of time.
Prediction: [{'label': 'NEGATIVE', 'score': 0.9997791647911072}]

Text: Not bad, but could be better.
Prediction: [{'label': 'NEGATIVE', 'score': 0.9467726945877075}]


### Выводы

После применения post-training dynamic quantization:

1. **Размер модели** уменьшился примерно в 2 раза (с ~268 MB до ~138 MB)
2. **Accurcay**: Потеря точности 1.5%
3. **Ускорение** инференса на CPU - 1.72x
